<a href="https://colab.research.google.com/github/mtryptnkr-study/Rock-vs-Mine-Prediction/blob/main/Rock_vs_Mine_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Context

* Consider there is a submarine and a war going on between two countries. Submarine of a country is going to another country. The enemy country has placed some mines under water. There can also be rocks in the ocean. Our job is to make a model to identify if there's a rock or a mine underneath the submarine.

* Submarine sends a sound signal which is known as Sonar and detects the object details.

* In real cases, the data has been collected by sending and receiving the sonar signals to and from the metal and a rock. In laboratory setup, the metal cylinder represents the mine and rock is the ground.

# Methodology

* Collection of Data
* Pre-process the Data
* Train-test split
* Feed it to a ML model
  * Used Logistic Regression model
* Train our model
* Test our model

## Importing the dependencies

In [131]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

## Uploading the dataset

In [132]:
# loading the dataset into DataFrame

sonar_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Machine Learning Projects/Sonar - Rock vs Mine Prediction/Sonar data.csv', header = None)

## Exploring the Data

In [133]:
sonar_data.head()

,0,1,2,3,4,5,6,7,8,9,...,51,52,53,54,55,56,57,58,59,60
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


### Observations
* Final column has R and M which represents Rock and Mine respectively.


In [134]:
sonar_data.shape

(208, 61)

In [135]:
sonar_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

### Statistical Measures

In [136]:
sonar_data.describe()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


* ```value_counts()``` is used to get the number of instances of a category.
* 60 represents the last column on which we have to focus on.

In [137]:
sonar_data[60].value_counts()

,count
60,
M,111
R,97


* As both the categories have almost equal instances, the accuracy of model is not compromised much.
* This is a balanced dataset.
* More the data, the better is the accuracy.

### Grouping the data for Mine and Rock

In [138]:
sonar_data.groupby(60).mean()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
60,,,,,,,,,,,,,,,,,,,,,
M,0.034989,0.045544,0.050720,0.064768,0.086715,0.111864,0.128359,0.149832,0.213492,0.251022,...,0.019352,0.016014,0.011643,0.012185,0.009923,0.008914,0.007825,0.009060,0.008695,0.006930
R,0.022498,0.030303,0.035951,0.041447,0.062028,0.096224,0.114180,0.117596,0.137392,0.159325,...,0.012311,0.010453,0.009640,0.009518,0.008567,0.007430,0.007814,0.006677,0.007078,0.006024


* ```groupby()``` is used to group the columns on which we have to focus our study on.
* We found that for each category there is a significant difference of the mean values for all the features.

## Separating the Data and Labels

In [139]:
X = sonar_data.drop(columns = 60, axis = 1)
y = sonar_data[60]

In [140]:
X.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0232,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0125,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0033,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0241,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0156,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094


In [141]:
y.head()

,60
0,R
1,R
2,R
3,R
4,R


## Splitting the data into Training and Test data

In [142]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, stratify = y, random_state = 42)


* ```test_size = 0.1``` means 10% of the dataset is test data and rest 90% dataset is training data.
* ```stratify``` ensure the equal splitting of R and M for test and train. It prevents skewing.
* ```random_state``` is used for specifying the splitting pattern.

In [143]:
print(X.shape, X_train.shape, X_test.shape)

(208, 60) (187, 60) (21, 60)


## Training the Model

### Logistic Regression Model
* This model works the best in the binary classification problem. The problem is to predict if the detected surface is a rock or a mine.
* This is a Supervised Learning model.

In [144]:
model = LogisticRegression()

In [145]:
model.fit(X_train, y_train)

LogisticRegression()

## Model Evaluation

In [146]:
# predicting the training data
X_train_prediction = model.predict(X_train)

In [147]:
# calculating the accuracy_score
training_data_accuracy = accuracy_score(X_train_prediction, y_train)
print(f'Accuracy on training data: {training_data_accuracy}')

Accuracy on training data: 0.8235294117647058


In [148]:
# predicting the testdata
X_test_prediction = model.predict(X_test)


In [149]:
# calculating the accuracy_score
testing_data_accuracy = accuracy_score(X_test_prediction, y_test)
print(f'Accuracy on testing data: {testing_data_accuracy}')

Accuracy on testing data: 0.8571428571428571


## Random Forest Algorithm
* This model can capture non-linear relationships which the logistic regression cannot capture.

In [150]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

In [151]:
# n_estimators is the number of trees in the forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

In [152]:
# Fitting the model
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

### Evaluating the model

In [153]:
# predicting the training data
X_train_prediction_rf = rf_model.predict(X_train)

In [154]:
# calculating the accuracy_score for training data
training_data_accuracy_rf = accuracy_score(X_train_prediction_rf, y_train)
print(f'Accuracy on training data: {training_data_accuracy_rf}')

Accuracy on training data: 1.0


In [155]:
# predicting the testing data
X_test_prediction_rf = rf_model.predict(X_test)

In [156]:
# calculating the accuracy_score for testing data
testing_data_accuracy_rf = accuracy_score(X_test_prediction_rf, y_test)
print(f'Accuracy on testing data: {testing_data_accuracy_rf}')

Accuracy on testing data: 0.9047619047619048


In [157]:
# making predictions
y_rf_pred = rf_model.predict(X_test)

In [158]:
print(f'Accuracy of Random Forest Classifier is: {accuracy_score(y_test, y_rf_pred)}')

Accuracy of Random Forest Classifier is: 0.9047619047619048


In [159]:
print("Classification Report:\n", classification_report(y_test, y_rf_pred))

Classification Report:
               precision    recall  f1-score   support

           M       0.85      1.00      0.92        11
           R       1.00      0.80      0.89        10

    accuracy                           0.90        21
   macro avg       0.92      0.90      0.90        21
weighted avg       0.92      0.90      0.90        21



### Random Forest using Hyperparameter Tuning

In [160]:
# Trying these parameters to prevent overfitting
rf_model_hp = RandomForestClassifier(
    n_estimators=100,
    max_depth=10,        # Limits how deep the trees grow
    min_samples_split=5, # Requires more data to make a "decision" branch
    max_features='sqrt', # Limits how many frequencies each tree looks at
    random_state=42
)


In [161]:
# Fitting the model
rf_model_hp.fit(X_train, y_train)

RandomForestClassifier(max_depth=10, min_samples_split=5, random_state=42)

In [162]:
# making predictions
y_pred_hp = rf_model_hp.predict(X_test)

In [163]:
print(f'Accuracy of Random Forest Classifier after hyperparameter tuning is: {accuracy_score(y_test, y_pred_hp)}')

Accuracy of Random Forest Classifier after hyperparameter tuning is: 0.8571428571428571


In [164]:
print("Classification Report:\n", classification_report(y_test, y_pred_hp))

Classification Report:
               precision    recall  f1-score   support

           M       0.79      1.00      0.88        11
           R       1.00      0.70      0.82        10

    accuracy                           0.86        21
   macro avg       0.89      0.85      0.85        21
weighted avg       0.89      0.86      0.85        21



## Observations:
* The accuracy is reducing due to the 'small data effect'.
* The sonar data contains just 208 rows of data. This is a small dataset.

## Making a Predictive System

In [165]:
input_data = (0.0346,0.0509,0.0079,0.0243,0.0432,0.0735,0.0938,0.1134,0.1228,0.1508,0.1809,0.2390,0.2947,0.2866,0.4010,0.5325,0.5486,0.5823,0.6041,0.6749,0.7084,0.7890,0.9284,0.9781,0.9738,1.0000,0.9702,0.9956,0.8235,0.6020,0.5342,0.4867,0.3526,0.1566,0.0946,0.1613,0.2824,0.3390,0.3019,0.2945,0.2978,0.2676,0.2055,0.2069,0.1625,0.1216,0.1013,0.0744,0.0386,0.0050,0.0146,0.0040,0.0122,0.0107,0.0112,0.0102,0.0052,0.0024,0.0079,0.0031)

In [166]:
# changing the input_data to a numpy array -> faster calculations
input_data_as_numpy_array = np.asarray(input_data)

### Logistic Regressor

In [167]:
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, 60)

prediction = model.predict(input_data_reshaped) # output is a list
print(prediction)

if (prediction[0] == 'R'):
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine


### Random Forest Predictor

In [168]:
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, 60)

prediction = rf_model.predict(input_data_reshaped) # output is a list
print(prediction)

if (prediction[0] == 'R'):
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine


### Hypertuned RF Predictor

In [169]:
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, 60)

prediction = rf_model_hp.predict(input_data_reshaped) # output is a list
print(prediction)

if (prediction[0] == 'R'):
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine


# Advanced Prediction

## Applying SMOTE (Synthetic Minority Over-sampling Technique)

In [170]:
from imblearn.over_sampling import SMOTE
from collections import Counter

# Apply SMOTE to the Training data ONLY
# k_neighbors=5 is default; you can lower it if a class is extremely small
smote = SMOTE(random_state=42, k_neighbors=5)
X_train_reshaped, y_train_reshaped = smote.fit_resample(X_train, y_train)

print(f"Resampled training shape: {Counter(y_train_reshaped)}")


Resampled training shape: Counter({'M': 100, 'R': 100})


In [171]:
# Now train your model on the reshaped data

rf_model_2 = RandomForestClassifier()
rf_model_2.fit(X_train_reshaped, y_train_reshaped)

RandomForestClassifier()

In [172]:
# making predictions
y_rf_pred_2 = rf_model_2.predict(X_test)

In [173]:
print(f'Accuracy of Random Forest Classifier after applying SMOTE is: {accuracy_score(y_test, y_rf_pred_2)}')

Accuracy of Random Forest Classifier after applying SMOTE is: 0.9047619047619048


## GridSearchCV

In [174]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from imblearn.pipeline import Pipeline

In [175]:
# Define the pipeline: First SMOTE, then Classifier
model_pipeline = Pipeline([
    ('smote', SMOTE(random_state=42)),
    ('classifier', RandomForestClassifier(random_state=42))
])

In [176]:
from scipy.stats import randint

param_dist = {
    'classifier__n_estimators': randint(100, 500),
    'classifier__max_depth': [None, 3, 5, 7, 10],
    'classifier__min_samples_split': randint(2, 20),
    'classifier__min_samples_leaf': randint(1, 10),
    'classifier__max_features': ['sqrt', 'log2', None]
}

# n_iter=50 means it will try 50 random combinations
random_search = RandomizedSearchCV(model_pipeline, param_distributions=param_dist,
                                   n_iter=50, cv=5, scoring='accuracy',
                                   random_state=42, n_jobs=-1)
random_search.fit(X_train, y_train)

print(f"Best RandomSearch Params: {random_search.best_params_}")
print(f"Best RandomSearch Score: {random_search.best_score_}")


Best RandomSearch Params: {'classifier__max_depth': 7, 'classifier__max_features': 'log2', 'classifier__min_samples_leaf': 1, 'classifier__min_samples_split': 17, 'classifier__n_estimators': 161}
Best RandomSearch Score: 0.7965860597439545


In [177]:
# Note the 'classifier__' prefix to tell the pipeline which step the params belong to
param_grid = {
    'classifier__n_estimators': [100, 200, 300],
    'classifier__max_depth': [None, 5, 10],
    'classifier__min_samples_split': [2, 5],
    'classifier__max_features': ['sqrt', 'log2']
}

grid_search = GridSearchCV(model_pipeline, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)

print(f"Best GridSearch Params: {grid_search.best_params_}")
print(f"Best GridSearch Score: {grid_search.best_score_}")

Best GridSearch Params: {'classifier__max_depth': None, 'classifier__max_features': 'log2', 'classifier__min_samples_split': 5, 'classifier__n_estimators': 300}
Best GridSearch Score: 0.8071123755334282


In [178]:
# Extract the best model
# final_model = random_search.best_estimator_

# If GridSearchCV, use:
final_model = grid_search.best_estimator_

final_model.fit(X_train, y_train)


Pipeline(steps=[('smote', SMOTE(random_state=42)),
                ('classifier',
                 RandomForestClassifier(max_features='log2',
                                        min_samples_split=5, n_estimators=300,
                                        random_state=42))])

In [179]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions
final_preds = final_model.predict(X_test)

# Check the results
print(f"Final Optimized Accuracy: {accuracy_score(y_test, final_preds)}")
print("\n--- Detailed Report ---")
print(classification_report(y_test, final_preds))


Final Optimized Accuracy: 0.8571428571428571

--- Detailed Report ---
              precision    recall  f1-score   support

           M       0.79      1.00      0.88        11
           R       1.00      0.70      0.82        10

    accuracy                           0.86        21
   macro avg       0.89      0.85      0.85        21
weighted avg       0.89      0.86      0.85        21



In [180]:
# See exactly which ones it got wrong (Rock vs Mine)
print("\n--- Confusion Matrix ---")
print(confusion_matrix(y_test, final_preds))


--- Confusion Matrix ---
[[11  0]
 [ 3  7]]


In [181]:
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, 60)

prediction = final_model.predict(input_data_reshaped) # output is a list
print(prediction)

if (prediction[0] == 'R'):
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine
